src/full_client.toit

```diff
@@ -816,17 +953,96 @@ class FullClient:
+    else if return_code == ConnAckPacket.IDENTIFIER_REJECTED:
+      refused_reason = "IDENTIFIER_REJECTED"
+    else if return_code == ConnAckPacket.SERVER_UNAVAILABLE:
+      refused_reason = "SERVER_UNAVAILABLE"
+    else if return_code == ConnAckPacket.BAD_USERNAME_OR_PASSWORD:
+      refused_reason = "BAD_USERNAME_OR_PASSWORD"
+    else if return_code == ConnAckPacket.NOT_AUTHORIZED:
+      refused_reason = "NOT_AUTHORIZED"
+    return refused_reason
 
   /**
   Connects (or reconnects) to the broker.
 
   Uses the reconnection_strategy to do the actual connecting.
   */
   reconnect_ --is_initial_connection/bool -> none:
-    throw "UNIMPLEMENTED"  // follow-up PR.
+    assert: not connection_ or not connection_.is_alive
+    old_connection := connection_
+
+    connecting_.do:
+      // Check that nobody else reconnected while we took the lock.
+      if connection_ != old_connection: return
+
+      if not connection_:
+        assert: is_initial_connection
+        connection_ = Connection_ transport_ --keep_alive=session_.options.keep_alive
+
+      try:
+        reconnection_strategy_.connect transport_
+            --is_initial_connection = is_initial_connection
+            --reconnect_transport = :
+              transport_.reconnect
+              connection_ = Connection_ transport_ --keep_alive=session_.options.keep_alive
+            --send_connect = :
+              packet := ConnectPacket session_.options.client_id
+                  --clean_session=session_.options.clean_session
+                  --username=session_.options.username
+                  --password=session_.options.password
+                  --keep_alive=session_.options.keep_alive
+                  --last_will=session_.options.last_will
+              connection_.write packet
+            --receive_connect_ack = :
+              response := connection_.read
+              if not response: throw "CONNECTION_CLOSED"
+              ack := (response as ConnAckPacket)
+              return_code := ack.return_code
+              if return_code != 0:
+                refused_reason := refused_reason_for_return_code_ return_code
+                connection_.close --reason=refused_reason
+                // No need to retry.
+                close --force
+                throw refused_reason
+              ack.session_present
+            --disconnect = :
+              connection_.write DisconnectPacket
+      finally: | is_exception exception |
+        if is_exception:
+          close --force
+
+      // If we are here, then the reconnection succeeded.
+
+      // Make sure the connection sends pings so the broker doesn't drop us.
+      connection_.keep_alive --background
+
+      // Resend the pending messages.
+      session_.do --pending: | packet_id/int persistent_id/int |
+        // The persistence store is allowed to decide not to resend packets.
+        persistence_store.get persistent_id
+          --if_absent=: session_.remove_pending packet_id
```

Comment:
This probably shows that it isn't super useful to get passed the `persistent_id` as the argument to the `if_absent` block.

Reply:
Changed the API in the previous PR.